### PR TITLE
Added smoothed softmax gradients

### PR DESCRIPTION
**Pull request by KellenSunderland**
MShadow modifications to support gradient smoothing for the softmax operator.

@larroy @fhieber @tdomhan

### CODE REVIEW

File: mshadow/cuda/tensor_gpu-inl.cuh [hunk 1]
```diff
@@ -256,6 +256,27 @@ __global__ void SoftmaxGradKernel(DstPlan dst, SrcPlan1 src, SrcPlan2 label, ind
 }
 
 template<int x_bits, typename DType, typename DstPlan, typename SrcPlan1, typename SrcPlan2>
+__global__ void SmoothSoftmaxGradKernel(DstPlan dst, SrcPlan1 src, SrcPlan2 label, index_t xmax,
+                                        float alpha) {
+  const unsigned x_size = 1 << x_bits;
+  const int y = blockIdx.x;
+  const int k = static_cast<int>(label.Eval(0, y));
+  const float smooth_grad = (alpha / (xmax - 1));
```
Comment (tdomhan) [inline mshadow/cuda/tensor_gpu-inl.cuh]: is `xmax` the number of classes?

**Review by tdomhan — APPROVED** [review]
from my limited mshadow understanding this looks good. There's a lot of duplication, though.

File: mshadow/cuda/tensor_gpu-inl.cuh (hunk 1 above)
Comment (KellenSunderland) [reply mshadow/cuda/tensor_gpu-inl.cuh]: Yep.  number of classes / vocab size.

Comment (larroy) [reply mshadow/cuda/tensor_gpu-inl.cuh]: I would add a comment

Comment (KellenSunderland) [reply mshadow/cuda/tensor_gpu-inl.cuh]: Will do.  

The copy and paste here is a little too much imo.  I was thinking about refactoring and at the same time making the ignore_label argument a index_t.  I'm leaning towards doing this in a separate PR though, what do you think?

Comment (KellenSunderland) [reply mshadow/cuda/tensor_gpu-inl.cuh]: comment added.

Comment (tdomhan) [reply mshadow/cuda/tensor_gpu-inl.cuh]: separate PR sounds fine.

File: mshadow/tensor_cpu-inl.h [hunk 2]
```diff
@@ -302,6 +302,26 @@ inline void SoftmaxGrad(Tensor<cpu, 2, DType> dst,
 }
 
 template<typename DType>
+inline void SmoothSoftmaxGrad(Tensor<cpu, 2, DType> dst,
+                        const Tensor<cpu, 2, DType> &src,
+                        const Tensor<cpu, 1, DType> &label,
+                        const float alpha) {
+  const float smooth_grad = (alpha / (dst.size(1) - 1));
+#pragma omp parallel for
+  for (openmp_index_t y = 0; y < dst.size(0); ++y) {
```
Comment (larroy) [inline mshadow/tensor_cpu-inl.h:311]: I would change this for index_t which is the appropiate type.

**Pull request closed without merging**